### PR TITLE
Enhance the azcopy options configuration scope

### PR DESCRIPTION
**Pull request by abhi18av**
This PR builds upon the available settings with the core `azcopy` tool which is used for uploads and downloads in the context of Azure Blob storage service.

### CODE REVIEW

**Comment by abhi18av** [comment]
@pditommaso this PR is ready from my side - but depends on a sister PR https://github.com/nextflow-io/azcopy-tool/pull/1 where I've mentioned the major benefit of the update since [we source the `azcopy` tool](https://github.com/nextflow-io/nextflow/blob/master/plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzBatchService.groovy#L88) from that repo.

**Comment by abhi18av** [comment]
**Quick update:** 

I've had to remove the options which are not compatible with the older `10.8.0` version of `azcopy` which we use and now the behavior is as expected on the Azure side 


The following config with custom azcopy opts are reflected on the generated `.command.run` files ✅ 

```
      azcopy {
          blockSize        = "17"
          blobTier         = "Cool"
          putMD5           = true
          overwrite        = false
          checkMD5         = "NoCheck"
          copyToolInstallMode= 'task'
      }


```



The only problem remains to be solved is the last test in `AzBashLibTest` 

https://github.com/nextflow-io/nextflow/blob/db89af1e4a5f2021972a2bdf2e0bb2aed997245e/plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy#L163

**Comment by abhi18av** [comment]
The unit tests for the code and the functional test (via rnaseq-nf` are passing now 

```console
(base) bash-5.0$ ../../nextflow/launch.sh run main.nf -profile azb 
N E X T F L O W  ~  version 22.08.0-edge
Launching `main.nf` [fervent_descartes] DSL2 - revision: 4ba66eb0c8
 R N A S E Q - N F   P I P E L I N E
 ===================================
 transcriptome: /Users/eklavya/projects/code/nextflow_on_azure/rnaseq-nf/data/ggal/ggal_1_48850000_49020000.Ggal71.500bpflank.fa
 reads        : /Users/eklavya/projects/code/nextflow_on_azure/rnaseq-nf/data/ggal/ggal_gut_{1,2}.fq
 outdir       : az://batch-jobs/rnaseq-nf-publishdir
 
Uploading local `bin` scripts folder to az://batch-jobs/rnaseq-nf-workdir/tmp/7e/03ec1928a1e8327db898dd00687297/bin
[e3/17ed0c] process > RNASEQ:INDEX (ggal_1_48850000_49020000) [100%] 1 of 1 ✔
[7e/54d1f2] process > RNASEQ:FASTQC (FASTQC on ggal_gut)      [100%] 1 of 1 ✔
[5d/3cbff4] process > RNASEQ:QUANT (ggal_gut)                 [100%] 1 of 1 ✔
[04/c40c43] process > MULTIQC                                 [100%] 1 of 1 ✔
Staging foreign file: /Users/eklavya/projects/code/nextflow_on_azure/rnaseq-nf/multiqc

Done! Open the following report in your browser --> az://batch-jobs/rnaseq-nf-publishdir/multiqc_report.html

Completed at: 05-Aug-2022 11:26:07
Duration    : 1m 37s
CPU hours   : (a few seconds)
Succeeded   : 4


```

File: plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy [hunk 1]
```diff
@@ -26,22 +26,73 @@ import groovy.transform.CompileStatic
...
+    //-----------------------------------------------------
+    // Default values for azcopy CLI environment variables
+    //-----------------------------------------------------
+
+    /* DISABLED TILL WE UPGRADE TO azcopy v10.16.0
+    //Set time (in minutes) for how long AzCopy should try to upload files for each request before AzCopy times out.
+    static public final String DEFAULT_AZCOPY_REQUEST_TRY_TIMEOUT = "4"
+    String requestTryTimeout
+    */
+
```
Comment (pditommaso) [inline plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy:38]: if useless, let's remove it

File: plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy [hunk 2]
```diff
@@ -26,22 +26,73 @@ import groovy.transform.CompileStatic
...
+    static public final String DEFAULT_BLOCK_SIZE = "4"
     String blockSize
+
+    //Upload block blob to Azure Storage using this blob tier. (azcopy default: "None")
+    static public final String DEFAULT_BLOB_TIER = "None" // hot (None) | cool | archive
     String blobTier
 
+    /* DISABLED TILL WE UPGRADE TO azcopy v10.16.0
+    //Define the output verbosity (azcopy default: "default").
+    //Set the default to quiet to avoid azcopy logs within the .command.log files
```
Comment (pditommaso) [inline plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy:54]: we can add it later

File: plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy [hunk 3]
```diff
@@ -26,22 +26,73 @@ import groovy.transform.CompileStatic
...
+
         this.blockSize = config.blockSize ?: DEFAULT_BLOCK_SIZE
         this.blobTier = config.blobTier ?: DEFAULT_BLOB_TIER
+        this.putMD5 = config.putMD5 ?: DEFAULT_PUT_MD5
+        this.checkMD5 = config.checkMD5 ?: DEFAULT_CHECK_MD5
+        this.overwrite = config.overwrite ?: DEFAULT_OVERWRITE
+        //this.requestTryTimeout = config.requestTryTimeout ?: DEFAULT_AZCOPY_REQUEST_TRY_TIMEOUT
+        //this.outputLevel = config.outputLevel ?: DEFAULT_OUTPUT_LEVEL
+
     }
```
Comment (pditommaso) [inline plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy:96]: Let's remove it

File: plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy [hunk 4]
```diff
@@ -26,22 +26,73 @@ import groovy.transform.CompileStatic
...
+        assert config != null
+
         this.blockSize = config.blockSize ?: DEFAULT_BLOCK_SIZE
         this.blobTier = config.blobTier ?: DEFAULT_BLOB_TIER
+        this.putMD5 = config.putMD5 ?: DEFAULT_PUT_MD5
+        this.checkMD5 = config.checkMD5 ?: DEFAULT_CHECK_MD5
+        this.overwrite = config.overwrite ?: DEFAULT_OVERWRITE
+        //this.requestTryTimeout = config.requestTryTimeout ?: DEFAULT_AZCOPY_REQUEST_TRY_TIMEOUT
+        //this.outputLevel = config.outputLevel ?: DEFAULT_OUTPUT_LEVEL
+
```
Comment (pditommaso) [inline plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzCopyOpts.groovy:95]: Let's remove it

File: plugins/nf-azure/src/main/nextflow/cloud/azure/file/AzBashLib.groovy [hunk 5]
```diff
@@ -31,30 +31,67 @@ class AzBashLib extends BashFunLib<AzBashLib> {
...
         """
-        # custom env variables used for azcopy opts
-        export AZCOPY_BLOCK_SIZE_MB=${blockSize}
-        export AZCOPY_BLOCK_BLOB_TIER=${blobTier}
+        # Env variables used for azcopy opts
+        export _AZCOPY_BLOCK_SIZE_MB=${blockSize}
+        export _AZCOPY_BLOCK_BLOB_TIER=${blobTier}
+        export _AZCOPY_PUT_MD5=${putMD5}
+        export _AZCOPY_CHECK_MD5=${checkMD5}
+        export _AZCOPY_OVERWRITE=${overwrite}
```
Comment (pditommaso) [inline plugins/nf-azure/src/main/nextflow/cloud/azure/file/AzBashLib.groovy:91]: Why they are prefixed with `_` ?

Comment (abhi18av) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/file/AzBashLib.groovy:91]: @pditommaso , I did this to distinguish the custom env vars with the actual env vars such as `AZCOPY_REQUEST_TRY_TIMEOUT` which we had to remove now and would be a part of next iteration.

**Comment by abhi18av** [comment]
@pditommaso , thanks for the review. I have accommodated the change requests now.

File: plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy [hunk 6]
```diff
@@ -44,20 +50,20 @@ class AzBashLibTest extends Specification {
...
-                    azcopy cp "$source/*?$AZ_SAS" "$target" --recursive >/dev/null || {
+                    azcopy cp "$source/*?$AZ_SAS" "$target" --recursive --check-md5 $_AZCOPY_CHECK_MD5 --overwrite $_AZCOPY_OVERWRITE >/dev/null || {
                         rm -rf $target
                         >&2 echo "Unable to download path: $source"
                         exit 1
                     }
                 }
             }
-            '''.stripIndent()
+            '''.stripIndent().replaceAll(' ', '')
```
Comment (pditommaso) [inline plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy:60]: Why there's the need for `.replaceAll(' ', '')`? never needed this before

File: docs/azure.rst [hunk 7]
```diff
@@ -353,6 +353,11 @@ The following configuration options are available:
 ============================================== =================
 Name                                           Description
 ============================================== =================
+azure.azcopy.blockSize                          The block size (in MiB) when uploading to Azure Storage, and downloading from Azure Storage (default: , requires ``nf-azure@0.13.1``).
+azure.azcopy.blobTier                           The blob tier to be used while uploading to Azure Storage (default: , requires ``nf-azure@0.13.1``).
+azure.azcopy.putMD5                             Create an MD5 hash of each file during upload, and save the md5sum hash as the Content-MD5 property of the destination blob or file. (default: false)
+azure.azcopy.checkMD5                           Specifies how strictly MD5 hashes should be validated when downloading data. (default 'FailIfDifferent')
+azure.azcopy.overwrite                          Overwrite the conflicting files and blobs at the destination if this flag is set to true. (default: "false")
```
Comment (pditommaso) [inline docs/azure.rst:360]: is it really needed to give access to this attribute? the task should decide if overwrite the files or not

File: docs/azure.rst [hunk 8]
```diff
@@ -353,6 +353,11 @@ The following configuration options are available:
 ============================================== =================
 Name                                           Description
 ============================================== =================
+azure.azcopy.blockSize                          The block size (in MiB) when uploading to Azure Storage, and downloading from Azure Storage (default: , requires ``nf-azure@0.13.1``).
+azure.azcopy.blobTier                           The blob tier to be used while uploading to Azure Storage (default: , requires ``nf-azure@0.13.1``).
```
Comment (pditommaso) [inline docs/azure.rst:357]: `default: ` ?

File: docs/azure.rst [hunk 9]
```diff
@@ -353,6 +353,11 @@ The following configuration options are available:
...
+azure.azcopy.putMD5                             Create an MD5 hash of each file during upload, and save the md5sum hash as the Content-MD5 property of the destination blob or file. (default: false)
+azure.azcopy.checkMD5                           Specifies how strictly MD5 hashes should be validated when downloading data. (default 'FailIfDifferent')
```
Comment (pditommaso) [inline docs/azure.rst:359]: Any reason to not make this the default?

Comment (abhi18av) [reply docs/azure.rst:359]: Paolo, this property is just surfacing the `azcopy` default of "FailIfDifferent" for information purpose. The users can opt to change if necessary.

File: plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy (hunk 6 above)
Comment (abhi18av) [reply plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy:60]: Ah, this is the problem related to the large string comparisons while testing, Jorge lent his experience here and this tweak makes Spock happy with the string comparison.

Comment (jorgeaguileraseqera) [reply plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy:60]: hahaha, well in fact was a trick to use during dev in order to know where the difference comes but once solved and test is green can be removed

File: docs/azure.rst (hunk 7 above)
Comment (abhi18av) [reply docs/azure.rst:360]: The use-case I had in mind was that if a user completes the analysis and the results are uploaded (without the MD5 sum) - they could use this parameter to force the uploads with checksums.

> the task should decide if overwrite the files or not

Paolo, do you mean the `process.publishDir` mechanism here?

Comment (pditommaso) [reply docs/azure.rst:360]: No, I mean the md5 checksum can be turned on for any uploaded. I don't see any value giving it as an option to the user

File: plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy (hunk 6 above)
Comment (abhi18av) [reply plugins/nf-azure/src/test/nextflow/cloud/azure/file/AzBashLibTest.groovy:60]: Ah, I thought this was meant for the long run!

With the use of `replaceAll` in first and second tests, both pass. Okay, the first test doesn't pass without `replaceAll`, but for the second I've removed this.

File: docs/azure.rst (hunk 7 above)
Comment (abhi18av) [reply docs/azure.rst:360]: Paolo, circling back on this one - could you please confirm whether the option in question is `putMD5` or `checkMD5`? 

I'm thinking `putMD5` right? If so, I can accommodate this and make a push.

Comment (pditommaso) [reply docs/azure.rst:360]: Actually, both of them, would not be better just to enable by defualt

Comment (abhi18av) [reply docs/azure.rst:358]: Sure Paolo, I have accommodated this feedback and made the push.

**Comment by abhi18av** [comment]
@pditommaso , I've made the MD5 (check/put) behavior as the default and resolved all merge conflicts. 

Shall we merge ahead? 


The next round of changes would be done only after the `azcopy` tool has been upgraded via https://github.com/nextflow-io/azcopy-tool/pull/1 (WIP)

**Comment by abhi18av** [comment]
@pditommaso , shall we take this merge forward?

**Comment by abhi18av** [comment]
@pditommaso , shall we merge the PR with this edge release cycle?

**Review by pditommaso — APPROVED** [review]

**Comment by pditommaso** [comment]
Abi, there's the usual problem your PRs do not run the tests and I've tested locally and it fails

**Comment by abhi18av** [comment]
> 
Abi, there's the usual problem your PRs do not run the tests and I've tested locally and it fails

🤔 this is curious. Let me take a look why there's no automated testing here.

**Comment by pditommaso** [comment]
We lost the momentum on this. Closing it because I don't know the status of this PR

**Comment by abhi18av** [comment]
True 😞 


But quick question, do you think that this effort would be a useful addition since `fusion` is now available for Azure as well?

**Comment by pditommaso** [comment]
That's another good point. The ideal goal is to not rely anymore on azcopy and use instead fusion in the mid-term

**Comment by abhi18av** [comment]
I see, in that case let me know if there's a need for the retouches on this PR and then I can make the iteration and circle back. Otherwise, I agree, good opportunity to showcase `fusionfs` for this  👍

**Pull request closed without merging**